### PR TITLE
[Metrics] Implement cumulative aggregation for async Long and Double counters in MetricPoint

### LINKED ISSUE #7215: Unskip MetricApiTests.ObservableCounterSpatialAggregationTest
This test is currently marked as skipped:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/08f25bc5d7529bf0eb545f43f9dba288ebd6c089/test/OpenTelemetry.Tests/Metrics/MetricApiTests.cs#L880-L883

https://github.com/open-telemetry/opentelemetry-specification/issues/1874 is closed as completed but the test still fails.

We need to investigate the failure and take appropriate action (fix a bug, file an issue, delete the test, something else?).

### PR DESCRIPTION
**Pull request by nabutabu**
Fixes #7215 
Design discussion issue NA

## Changes

Use `InterlockedHelper.Add` in Update calls and reset aggregation values in `TakeSnapshot` using `Interlocked.Exchange(ref this.runningValue.AsLong, 0);`

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/MetricPoint/MetricPoint.cs
Comment (martincostello) [inline src/OpenTelemetry/Metrics/MetricPoint/MetricPoint.cs:1]: The changes seem/sound reasonable, but we need to make sure we have coverage for all of the changed branches.

File: src/OpenTelemetry/Metrics/MetricPoint/MetricPoint.cs [hunk 1]
```diff
@@ -375,6 +375,11 @@ internal void Update(long number)
                 }
 
             case AggregationType.LongSumIncomingCumulative:
+                {
+                    Interlocked.Add(ref this.runningValue.AsLong, number);
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricPoint/MetricPoint.cs]: This is not correct. If incoming is cumulative, then it is already the sum. We (i.e SDK) shouldn't add it.

**Review by cijothomas — CHANGES_REQUESTED** [review]
See
https://github.com/open-telemetry/opentelemetry-dotnet/pull/7227/changes#r3168921622

Let me know if I misunderstood the fix.

**Comment by nabutabu** [comment]
Fix was incorrect and the issue was that I did not take into account a couple of things which the failing tests also pointed out. Changing `Update`/`UpdateWithExemplar` was incorrect because each reader has its own `AggregatorStore` and so the same measurement stream feeds both views. Two measurements with different raw tags but the same filtered tags return the same index from `FindMetricAggregatorsCustomTag`. Then `UpdateLongMetricPoint` is called twice on that same index which then means that the second Exchange call overwrites the first.

Basically if a MetricPoint is in state CollectPending and is IsAsynchronous then the value sent by `UpdateLongCustomTags` is the existingRunningValue + newValue instead of just newValue.

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 2]
```diff
@@ -938,6 +938,12 @@ private void UpdateLongCustomTags(long value, ReadOnlySpan<KeyValuePair<string,
     {
         var index = this.FindMetricAggregatorsCustomTag(tags);
 
+        if (this.IsAsynchronous && this.aggType == AggregationType.LongSumIncomingCumulative && this.metricPoints[index].MetricPointStatus == MetricPointStatus.CollectPending)
+        {
+            this.UpdateLongMetricPoint(index, value + this.metricPoints[index].GetRunningValueLong(), tags);
```
Comment (martincostello) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs]: Same index is accessed twice here. I would split it into the first two conditions as one if, then get the item, then test the third condition, then use the local containing.

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 3]
```diff
@@ -985,6 +991,12 @@ private void UpdateDoubleCustomTags(double value, ReadOnlySpan<KeyValuePair<stri
     {
         var index = this.FindMetricAggregatorsCustomTag(tags);
 
+        if (this.IsAsynchronous && this.aggType == AggregationType.DoubleSumIncomingCumulative && this.metricPoints[index].MetricPointStatus == MetricPointStatus.CollectPending)
```
Comment (martincostello) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs]: Same here.

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 4]
```diff
@@ -938,6 +938,15 @@ private void UpdateLongCustomTags(long value, ReadOnlySpan<KeyValuePair<string,
     {
         var index = this.FindMetricAggregatorsCustomTag(tags);
 
+        if (this.IsAsynchronous && this.aggType == AggregationType.LongSumIncomingCumulative)
+        {
+            ref var metricPoint = ref this.metricPoints[index];
+            if (metricPoint.MetricPointStatus == MetricPointStatus.CollectPending)
+            {
+                value += metricPoint.GetRunningValueLong();
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:1003]: The idea of modifying the incoming value to achieve "spatial aggregation" is good - but it does create incorrect Exemplar values. 

It is not that bad - as Exemplars have limited utility in Observable instruments anyway and Exemplars are disabled also by-default for Observable. We should see if the fix can keep Exemplars correct as well. (Exemplar should get the original unmodified value, but Metricpoint should get the updated one..)

File: test/OpenTelemetry.Tests/Metrics/MetricApiTests.cs [hunk 5]
```diff
@@ -877,7 +877,7 @@ public void ObservableCounterWithTagsAggregationTest(bool exportDelta)
         ValidateMetricPointTags(tags3, metricPoint3.Tags);
     }
 
-    [Theory(Skip = "https://github.com/open-telemetry/opentelemetry-specification/issues/1874")]
+    [Theory]
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Metrics/MetricApiTests.cs:878]: I added just one test to prove that we don't do spatial-aggregation in this SDK. We need much more coverage when fixing the issue. Covering delta/cumulative, multiple cycles, various view combinations.

**Review by cijothomas — CHANGES_REQUESTED** [review]
The current iteration looks very promising (thanks! This was indeed a hard problem!). 
Requesting changes to address below comment
https://github.com/open-telemetry/opentelemetry-dotnet/pull/7227/changes#r3183554865

Also, we need to vastly improve test coverage when doing this fix- lot of potential edge cases. (I spotted only one about Exemplar value, but could be more)

File: src/OpenTelemetry/Metrics/AggregatorStore.cs (hunk 4 above)
Comment (nabutabu) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:1003]: Read more about Exemplars and this does complicate this issue a bit. Blurting out my understanding of what's happening for future me:

Exemplars are defined as: "access to the **raw measurement value**, time stamp when measurement was made, and **trace context**", which means each exemplar recording should be the raw reading instead of the cumulative value that the (non-exemplar) async cumulative instrument uses. 

This is important because each exemplar metric recording is also associated with a trace. If exemplars stored aggregated values, we'd lose the connection to individual traces for each recording. The exemplar's purpose is to preserve the context of specific measurements that contribute to the aggregate, which this current change loses. 

I'll come back when I have more on how we could fix this. 
- First thought is to add a parameter to `UpdateWithExemplar`  in `MetricPoint` -> "number" would be the cumulative value and the new parameter is the measured value.  
- MetricPoint does already have access to its own running value, maybe we could just subtract that when calling UpdateExemplar, this would need to happen before `Interlocked.Add` is called on the running value. Also would this work for every kind of instrument?

**Comment by nabutabu** [comment]
@cijothomas  I added as many tests as I could think of, but I am not sure if this is all that needs coverage. Would appreciate any guidance on any other nuances of the measurements that could potentially be affected!

**Comment by Kielek** [comment]
Asked Codex/GPT5.5 to check this PR.
Both of them marked as P1. Could you please double check this feedback?

# Review Notes and Unit Test Proposals for PR #7227

I think there are two correctness risks worth covering with targeted unit tests.

## Issue 1: Delta temporality is not correct when filtered streams change

The current fix spatially aggregates async cumulative counter values by updating the already-filtered `MetricPoint` with:

```text
existing running value + incoming measurement value
```

That works for simple cumulative export cases where the same underlying streams are reported every cycle. The problem is delta export. `MetricPoint.TakeSnapshot(outputDelta: true)` computes the delta from the collapsed aggregate:

```text
current collapsed aggregate - previous collapsed aggregate
```

That is not equivalent to summing per-stream deltas if one original measurement stream disappears or reappears between callbacks.

Example:

```text
Collection 1:
  A = 10
  B = 10
  Collapsed value = 20

Collection 2:
  A = 15
  B is absent
  Collapsed value = 15
```

If the SDK diffs collapsed values, delta export becomes:

```text
15 - 20 = -5
```

The expected delta is:

```text
(15 - 10) = 5
```

The absent stream should not subtract from the collapsed aggregate. Delta needs to be calculated per original measurement stream before spatial aggregation, or the implementation needs equivalent per-stream state.

### Suggested test: long observable counter, one stream disappears

Add a regression test for `ObservableCounter<long>` with delta temporality and a view that drops all tags.

Scenario:

```text
Collection 1:
  A = 10
  B = 10
  Expected exported delta = 20

Collection 2:
  A = 15
  B is absent
  Expected exported delta = 5
```

This should fail if the SDK diffs only the collapsed aggregate, because it would export `-5`.

### Suggested test: long observable counter, unchanged stream and missing stream

This is the smallest version of the same bug.

Scenario:

```text
Collection 1:
  A = 10
  B = 10
  Expected exported delta = 20

Collection 2:
  A = 10
  B is absent
  Expected exported delta = 0
```

An aggregate-diff implementation would export:

```text
10 - 20 = -10
```

The expected result is zero because the only stream still reported did not change.

### Suggested test: double observable counter, one stream disappears

Add the same coverage for `ObservableCounter<double>`.

Scenario:

```text
Collection 1:
  A = 10.5
  B = 10.5
  Expected exported delta = 21.0

Collection 2:
  A = 15.5
  B is absent
  Expected exported delta = 5.0
```

This covers the `UpdateDoubleCustomTags` path, which has the same aggregation risk as the long path.

### Suggested test: stream reappears after being absent

This verifies that per-stream state is preserved across multiple collection cycles.

Scenario:

```text
Collection 1:
  A = 10
  B = 10
  Expected exported delta = 20

Collection 2:
  A = 15
  B is absent
  Expected exported delta = 5

Collection 3:
  A = 20
  B = 25
  Expected exported delta = 20
```

Collection 3 should be:

```text
(20 - 15) + (25 - 10) = 20
```

This catches both disappearing and reappearing stream behavior under spatial aggregation.

## Issue 2: Custom-tag overflow can throw before the overflow metric point is used

`FindMetricAggregatorsCustomTag` can return `-1` when the metric point cardinality limit is exhausted. Before this PR, that `-1` was handled by `UpdateLongMetricPoint` or `UpdateDoubleMetricPoint`, which routes the measurement to the overflow metric point.

The new custom-tag async branches read the metric point before that existing negative-index handling runs:

```text
ref var metricPoint = ref this.metricPoints[index];
```

If `index` is `-1`, this throws before the overflow path can run. The outer `AggregatorStore.Update` catches the exception and records an SDK internal error/drop, so the existing overflow behavior is bypassed.

This applies to both:

```text
UpdateLongCustomTags
UpdateDoubleCustomTags
```

### Suggested test: long observable counter custom-tag overflow

Add a test that forces the custom-tag path and exhausts the cardinality limit.

Shape:

```text
Instrument:
  ObservableCounter<long>

View:
  TagKeys = ["keep"]

Measurements:
  Use enough distinct "keep" values to exhaust the configured metric point cardinality limit.

Assertions:
  ForceFlush does not throw.
  The existing overflow metric point behavior is used.
  The measurement is not dropped as an SDK internal error.
```

This guards against indexing `metricPoints[-1]` before `UpdateLongMetricPoint` can route the measurement to overflow.

**Comment by nabutabu** [comment]
The delta temporality behavior under spatial aggregation is a known open question for the spec and @cijothomas has already filed [open-telemetry/opentelemetry-specification#4861](https://github.com/open-telemetry/opentelemetry-specification/issues/4861) to track it.

The core issue is what delta to emit when a stream disappears or reappears:
- **Option 1 (delta from zero):** 
    Bounded memory, but incorrect if the underlying counter continued incrementing during the gap.
- **Option 2 (delta from last known value):** 
    Correct, but requires tracking all ever-seen attribute sets, which conflicts with cardinality limits and the SDK's existing reclaim machinery.

I think we should skip the four failing delta tests I added and reference the issue mentioned above:

```C#
ObservableCounterSpatialAggregationDelta_UnchangedStreamAndMissingStream_DeltaIsZero
ObservableCounterSpatialAggregationDelta_StreamDisappears_DeltaIsCorrect
ObservableCounterSpatialAggregationDelta_StreamReappearsAfterGap_DeltaIsCorrect
ObservableCounterSpatialAggregationDelta_Double_StreamDisappears_DeltaIsCorrect
```


Seems like we added the fix to remove `Skip` for one test and ended up adding 4 more tests that we need to skip haha

**Comment by nabutabu** [comment]
@cijothomas Could I get another review here whenever possible? Thank you :)
